### PR TITLE
feat(condo): added twilio and instasent adapters for sending SMS

### PR DESCRIPTION
**Pull request by paulo-rossy**

### CODE REVIEW

File: apps/condo/domains/notification/adapters/smsAdapter.js [hunk 1]
```diff
@@ -58,6 +66,122 @@ class SMSAdapter {
...
+                    'Accept': 'application/json',
+                },
+            }
+        )
+        return result.ok
+    }
+
+    async send ({ phone, message }, extendedParams = {}) {
+        const body = {
+            clientId: crypto.randomBytes(16).toString('hex'),
```
Comment (sitozzz) [inline apps/condo/domains/notification/adapters/smsAdapter.js:101]: Strange behaviour of client id generation for every single request

Comment (SavelevMatthew) [reply apps/condo/domains/notification/adapters/smsAdapter.js:101]: Maybe something `sender.fingerprint` can be used here?

File: apps/condo/domains/notification/adapters/smsAdapter.js [hunk 2]
```diff
@@ -58,6 +66,122 @@ class SMSAdapter {
...
+            from: 'CondoDevApi',
```
Comment (SavelevMatthew) [inline apps/condo/domains/notification/adapters/smsAdapter.js:102]: Should it be part of config obtained from `.env`?

File: apps/condo/domains/notification/adapters/smsAdapter.js [hunk 3]
```diff
@@ -58,6 +66,122 @@ class SMSAdapter {
...
+                },
+            }
+        )
+        return result.ok
+    }
+
+    async send ({ phone, message }, extendedParams = {}) {
+        const body = {
+            To: phone,
+            From: this.fromNumber,
```
Comment (SavelevMatthew) [inline apps/condo/domains/notification/adapters/smsAdapter.js:162]: What is this for? Is it like phone-2-phone message provider?

File: apps/condo/domains/notification/adapters/smsAdapter.js [hunk 4]
```diff
@@ -58,6 +66,122 @@ class SMSAdapter {
...
+            Body: message,
+            ...extendedParams,
+        }
+        const result = await fetch(
+            `https://api.twilio.com/2010-04-01/Accounts/${this.accountSid}/Messages.json`,
```
Comment (SavelevMatthew) [inline apps/condo/domains/notification/adapters/smsAdapter.js:167]: strange API url, why `2010` is here?

Comment (SavelevMatthew) [reply apps/condo/domains/notification/adapters/smsAdapter.js:167]: Ok, seems like I've got it. It's legit API url

File: apps/condo/domains/notification/adapters/smsAdapter.js [hunk 5]
```diff
@@ -58,6 +66,122 @@ class SMSAdapter {
...
+            {
+                method: 'POST',
+                headers: {
+                    'Authorization': 'Basic ' + Buffer.from(this.accountSid + ':' + this.authToken).toString('base64'),
+                    'Content-Type': 'application/x-www-form-urlencoded', 
+                },
+                body: Object.entries(body)
```
Comment (SavelevMatthew) [inline apps/condo/domains/notification/adapters/smsAdapter.js:174]: Why are you forming query-url in body? Looks strange

File: apps/condo/domains/notification/adapters/smsAdapter.js (hunk 3 above)
Comment (SavelevMatthew) [reply apps/condo/domains/notification/adapters/smsAdapter.js:162]: Read their docs, they're phone-2-phone...

File: apps/condo/domains/notification/adapters/smsAdapter.js (hunk 5 above)
Comment (SavelevMatthew) [reply apps/condo/domains/notification/adapters/smsAdapter.js:174]: I guess you should just be able to pass body object like this:
```suggestion
                body: JSON.stringify(body)
```

**Comment by SavelevMatthew** [comment]
Hi, thanks for your contribution to our open-source project 🙏. 

I've read through your changes and overall everything looks good, but I have a couple small nuances that I'd like to clarify or ask you to fix:

1. Why does instasent use a random string? If it's a client id, it would be better to use `sender.fingerprint`, which we can get in almost all our queries. Or if it's an API server identifier, why regenerate it?
2. I wrote a bit above about how to more efficiently generate body for form-urlencoded, it would be cool if you could fix that.
3. It seems like the sender name in instasent should be passed through env, so you can change it quickly if needed. Just like fromNumber in triwio
4. It might be worth moving the API-url to env parameters like in other providers to keep the parameters consistent

Otherwise everything is cool! I'll be happy to merge your PR once we've solved the issues above 🥳 🤝

File: apps/condo/domains/notification/adapters/smsAdapter.js [hunk 6]
```diff
@@ -58,6 +64,125 @@ class SMSAdapter {
     }
 }
 
+class InstasentSms {
+
+    isConfigured = false
```
Comment (sitozzz) [inline apps/condo/domains/notification/adapters/smsAdapter.js:69]: You don't really need to save value of `validateConfig` inside class instance. You can just call `validateConfig` inside constructor of your adapter class.
```suggestion
```

File: apps/condo/domains/notification/adapters/smsAdapter.js [hunk 7]
```diff
@@ -58,6 +64,125 @@ class SMSAdapter {
...
+        )
+        const json = await result.json()
+        const isOk =  result.ok
+        return [isOk, json]
+    }
+}
+
+class TwilioSms {
+
+    isConfigured = false
```
Comment (sitozzz) [inline apps/condo/domains/notification/adapters/smsAdapter.js:128]: Same as above
```suggestion
```

**Review by SavelevMatthew — APPROVED** [review]
Looks good! Seems like it's ready to merge after passing CI pipelines! 🔥 🥳

**Comment by sitozzz** [comment]
Anyway it looks pretty good to me. My comments are optional to fix. But one moment you should done - change PR origin from `open-condo-software:master` to `master`

**Review by sitozzz — APPROVED** [review]

**Comment by SavelevMatthew** [comment]
> Anyway it looks pretty good to me. My comments are optional to fix. But one moment you should done - change PR origin from `open-condo-software:master` to `master`

This notation is default when you try to merge from forks, if you click on `open-condo-software:master` you'll redirected to repo master branch

**Comment by sitozzz** [comment]
> > Anyway it looks pretty good to me. My comments are optional to fix. But one moment you should done - change PR origin from `open-condo-software:master` to `master`
> 
> This notation is default when you try to merge from forks, if you click on `open-condo-software:master` you'll redirected to repo master branch

Your truth)

**Comment by SavelevMatthew** [comment]
Hello again. It turns out that our CI system is not ready for PR from forks yet. We are ready to merge your PR, and we will try to fix CI soon. Thanks again for your contribution 🥳 🎈

**Pull request merged**